Generate html and json of supported css properties.

jrasanen opened the pull request:
Fixes #10196. Outputs html and json of supported css properties to `target/doc/` directory when deploying github-pages.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10208)

<!-- Reviewable:end -->

jrasanen committed:
Generate html and json of supported css properties.

Outputs html and json file of supported css properties to target/doc/
directory when deploying github-pages.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/list_properties.py, components/style/properties.html.mako
 * @simonsapin: components/style/list_properties.py, components/style/properties.html.mako

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

SimonSapin commented on components/style/properties.html.mako:
```diff
@@ -0,0 +1,67 @@
+    </section>
+
+    <script src="../jquery.js"></script>
+    <script>
```
Since this is a Mako template that has `properties` available, could Mako generate the HTML table directly, without JavaScript?

jrasanen replied:
@SimonSapin Hi! Fixed the template to use directly Mako instead of Javascript + JSON.

jrasanen committed:
HTML generation with Mako instead of JS

larsbergstrom commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/list_properties.py, line 37 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10208#-KDpVDktEomaTd5efEws:-KDpVDktEomaTd5efEwt:-1842596154) ([raw file](https://github.com/servo/servo/blob/b87d6e7615ef97e637e5d13d123cc50375bbe39f/components/style/list_properties.py#L37)):*
Would it be possible to restructure this so that we don't edit list_properties.py, but instead in `upload_docs.sh`:

1) Redirect the output of this script to a file (e.g., css-properties.json, as you named it here)

2) Run the mako stuff, possibly in another small python script that lives in the etc/ci directory, if that is required.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10208)*
<!-- Sent from Reviewable.io -->

jrasanen commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/list_properties.py, line 37 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10208#-KDpVDktEomaTd5efEws:-KDrTjLx1OylHyoNWUq_:-1117812075) ([raw file](https://github.com/servo/servo/blob/b87d6e7615ef97e637e5d13d123cc50375bbe39f/components/style/list_properties.py#L37)):*
Yes, agreed, having a script which just outputs the JSON is better since then its more generic. Also makes it easier just to pipe the output and change the end location of it in CI script.

My initial implementation was just a plain html reading the json with jquery (which seems to be included with the documentation). I could either do that again or do a python/shell script which generates the property table?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10208)*
<!-- Sent from Reviewable.io -->

jrasanen commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/list_properties.py, line 37 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10208#-KDpVDktEomaTd5efEws:-KDrbLkEwK8GRxJdqSma:-622020347) ([raw file](https://github.com/servo/servo/blob/b87d6e7615ef97e637e5d13d123cc50375bbe39f/components/style/list_properties.py#L37)):*
I think I got Simon Sapin's suggestion on Github wrong regarding generating the static html in that script. Sorry about that!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10208)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
@bors-servo r+

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[components/style/list_properties.py, line 37 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10208#-KDpVDktEomaTd5efEws:-KDxBrN1boK1IsIJjmLW:-722475860) ([raw file](https://github.com/servo/servo/blob/b87d6e7615ef97e637e5d13d123cc50375bbe39f/components/style/list_properties.py#L37)):*
What you have in this PR now is what I had in mind. Thanks!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10208#-:-KDxCR6EfCtdSCvrbSz8:1236647940)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit b87d6e7 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin b87d6e7615ef97e637e5d13d123cc50375bbe39f -->

SimonSapin commented:
@bors-servo r-

There’s an error on Travis-CI:

```
0.37s$ ./mach test-unit

Traceback (most recent call last):

  File "/home/travis/build/servo/servo/components/style/list_properties.py", line 32, in <module>

    with open(os.path.join(servo_doc_path, 'css-properties.json'), "w") as out_file:

IOError: [Errno 2] No such file or directory: '/home/travis/build/servo/servo/target/doc/servo/css-properties.json'
```

Please change the script so that it creates directories as needed if `target` does not exist. (Try using https://git-scm.com/docs/git-worktree if you have a build you don’t want to remove.)

jrasanen committed:
Fixed script to output the JSON and create doc directory.

jrasanen commented:
@SimonSapin good catch, modified script to create the directory and added the json printing back, I think `./mach test-unit` needs it too. Ran those tests locally too and it passed now.

SimonSapin commented:
Printing would make the CI logs for `upload_docs.sh` noisy. Could you rather change `test-unit` to read the JSON file after running the command?

jrasanen committed:
Changed mach test to use generated json file of properties

jrasanen commented:
@SimonSapin, modified mach test unit to run the script to generate JSON and read it afterwards

jrasanen committed:
Fixed a typo

SimonSapin commented:
Thanks!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 6de7228 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 6de7228945f9e9882ae5a8d45fd3ed40906057db -->

bors-servo commented:
:hourglass: Testing commit 6de7228 with merge b8ef3f9...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2521)

larsbergstrom commented:
Huh. New intermittent?
```
Tests with unexpected results:
  ▶ TIMEOUT [expected FAIL] /css21_dev/html4/clear-applies-to-008.htm
  │ 
  │ Shutting down the Constellation after generating an output file or exit flag specified
  │ thread 'Constellation' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } })', ../src/libcore/result.rs:746
  │ stack backtrace:
  │    1:     0x7f56144d7a40 - sys::backtrace::tracing::imp::write::h495bc662e480d01f2cv
  │    2:     0x7f56144dc89f - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.44522
  │    3:     0x7f56144dc518 - panicking::default_handler::h4e73712d1e927dfeH0z
  │    4:     0x7f56144c67dc - sys_common::unwind::begin_unwind_inner::h0e1be209f5e9e87dg2t
  │    5:     0x7f56144c7788 - sys_common::unwind::begin_unwind_fmt::h45dc2fac9b5e0d23m1t
  │    6:     0x7f56144d6d41 - rust_begin_unwind
  │    7:     0x7f56145142bf - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:     0x7f561382523e - result::unwrap_failed::h9898011520364998313
  │    9:     0x7f561384e389 - font_cache_thread::FontCacheThread::exit::ha2d0eb6658e647b1uMn
  │   10:     0x7f5612b3fa35 - constellation::Constellation<LTF, STF>::handle_exit::h10693835640492743256
  │   11:     0x7f5612b1cce6 - constellation::Constellation<LTF, STF>::handle_request::h6557854063750769946
  │   12:     0x7f5612b10f4c - sys_common::unwind::try::try_fn::h5641394920290239990
  │   13:     0x7f56144d6ccb - __rust_try
  │   14:     0x7f56144d6c5d - sys_common::unwind::inner_try::he9b0b11c6b5b4e7fiZt
  │   15:     0x7f5612b1240a - boxed::F.FnBox<A>::call_box::h4738823920327326506
  │   16:     0x7f56144dae69 - sys::thread::Thread::new::thread_start::hf1ffcc04a41608eb9Xy
  │   17:     0x7f561076c181 - start_thread
  │   18:     0x7f561028347c - __clone
  └   19:                0x0 - <unknown>

program finished with exit code 1
elapsedTime=640.913206
```

jdm commented:
@bors-servo: retry
* #8815

bors-servo commented:
:hourglass: Testing commit 6de7228 with merge d910ae1...

bors-servo commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1330)

jdm commented:
@bors-servo: retry
* appveyor weirdness

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:hourglass: Testing commit 6de7228 with merge 365f7e2...

bors-servo commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1361)

larsbergstrom commented:
@bors-servo retry

not sure what is going on here - it looks like the appveyor build is kicking off either before or after auto has moved to a change that does not contain this commit:
```
git clone -q --branch=auto https://github.com/servo/servo.git C:\projects\servo
git checkout -qf 365f7e2f685ce7d21bd37af1b9620467908f582f
fatal: reference is not a tree: 365f7e2f685ce7d21bd37af1b9620467908f582f
Command exited with code 128
```

jrasanen commented:
@larsbergstrom maybe squashing the commits & pushing would fix it?

bors-servo commented:
:hourglass: Testing commit 6de7228 with merge df73a18...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2200), [arm32](http://build.servo.org/builders/arm32/builds/132), [gonk](http://build.servo.org/builders/gonk/builds/2276), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2254), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2539), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/944), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2300), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2362), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1392)
